packages/eslint-plugin/src/stylex-enforce-extension.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow strict
+ */
+
+'use strict';
+
+import type { Node } from 'estree';
+
+/*:: import { Rule } from 'eslint'; */
+
+const stylexEnforceExtension = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Ensure that files exporting StyleX Vars using `stylex.defineVars` end with a specified extension (default `.stylex.jsx` or `.stylex.tsx`), and that files exporting other values must not use that extension. Mixed exports are not allowed. Users can define a custom extension using the `themeFileExtension` option.',
+      category: 'Possible Errors',
+      recommended: false,
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          themeFileExtension: {
+            type: 'string',
+            default: '.stylex.jsx',
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+  },
+  create(context: Rule.RuleContext): { ... } {
+    let hasStyleXVarsExports = false;
+    let hasOtherExports = false;
+    const fileName = context.getFilename();
+    const options = context.options[0] || {};
+    const themeFileExtension = options.themeFileExtension || '.stylex.jsx';
+    const themeTsxExtension = themeFileExtension.replace('.jsx', '.tsx');
+
+    return {
+      ExportNamedDeclaration(node: Node) {
+        if (node.declaration && node.declaration.declarations) {
+          node.declaration.declarations.forEach((declaration) => {
+            const init = declaration.init;
+
+            if (
+              init &&
+              init.callee &&
+              init.callee.object &&
+              init.callee.object.name === 'stylex' &&
+              init.callee.property &&
+              init.callee.property.name === 'defineVars'
+            ) {
+              hasStyleXVarsExports = true;
+            } else {
+              hasOtherExports = true;
+            }
+          });
+        }
```

Comment:
You can do something simple like this.

```suggestion
        if (node.declaration && node.declaration.declarations) {
          node.declaration.declarations.forEach((declaration) => {
            const init = declaration.init;

            if (
              init &&
              init.callee &&
              init.callee.object &&
              init.callee.object.name === 'stylex' &&
              init.callee.property &&
              init.callee.property.name === 'defineVars' &&
              !isStyleXFile
            ) {
              context.report({
                node: declaration.init,
                message: "...Must be in a .stylex.js file"
              });
            } else if (isStyleXFile) {
              context.report({
                node: declaration.init,
                message: "...Must NOT be in a .stylex.js file"
              });
            }
          });
        }
```